### ISSUE TITLE
False-positive `DTZ007`:  The use of `datetime.datetime.strptime()` without %z must be followed by `.replace(tzinfo=)`

### ISSUE THREAD

**Issue by actionless** [charliermarsh/ruff#1306]
```python
import datetime

DT_FORMAT = "%a, %d %b %Y %H:%M:%S %z"
file_data = "Fri, 23 Sep 2022 08:42:36 +0000"


parsed_date = datetime.datetime.strptime(
    file_data, DT_FORMAT
)
```

```console
$ ruff test_ruff_DTZ007.py
Found 1 error(s).
test_ruff_DTZ007.py:7:15: DTZ007 The use of `datetime.datetime.strptime()` without %z must be followed by `.replace(tzinfo=)`
```

**Comment by charliermarsh** [comment]
Probably a stretch to fix this right now since it relies on static analysis that's beyond Ruff's current capabilities (understanding that `DT_FORMAT` is assigned to a string containing `%z`).

**Comment by actionless** [comment]
yeah i was actually wondering how you going to address that in the feature, by reusing parts of RustPython to actually run some parts?

**Comment by andersk** [comment]
Just to dream a little bit here, I think there could be a future where we integrate with mypy or another type checker to enable type-aware lints.

In that world, one could write

```python
from typing import Final

DT_FORMAT: Final = "%a, %d %b %Y %H:%M:%S %z"
file_data = "Fri, 23 Sep 2022 08:42:36 +0000"

parsed_date = datetime.datetime.strptime(
    file_data, DT_FORMAT
)
```

The `Final` annotation ([PEP 591](https://peps.python.org/pep-0591/)) promises that the value won’t be reassigned, and allows the type checker to infer a type `Literal["%a, %d %b %Y %H:%M:%S %z"]` ([PEP 586](https://peps.python.org/pep-0586/)) for `DT_FORMAT`, which could be used in the Ruff rule. This kind of integration would enable us to improve the accuracy of many rules and enable a variety of new ones.

**Comment by charliermarsh** [comment]
As a starting point... we _could_ support `Final` specifically. Maybe that's slightly unprincipled though.

**Comment by gdub** [comment]
A similar case I ran across today was that grabbing only date-related items from a string, e.g.:
```python
from datetime import datetime
datetime.strptime("2022-01-01", "%Y-%m-%d").date()
```
...also results in a `DTZ007` false-positive:
```
$ ruff check --isolated --select DTZ strptime_test.py 
strptime_test.py:2:1: DTZ007 The use of `datetime.datetime.strptime()` without %z must be followed by `.replace(tzinfo=)` or `.astimezone()`
Found 1 error.
```

In a case like this where not dealing with time, the resulting date value is the same and is not dependent on a timezone. To detect this case, though, it seems that we'd have to recognize both that the format string contains no time-related items _and_ that there's a following conversion with `.date()`.

**Comment by charliermarsh** [comment]
@youknowone - We do have some of our abstractions in Ruff for implementing a semantic model -- we track scopes, bindings, etc. I want to spend some time improving the semantic model, so I'll take a look at what you have in RustPython as part of that exercise -- maybe there are things we can leverage!

**Comment by nachocab** [comment]
I'm also getting a false positive when using the uppercase version `%Z` (Time zone name: UTC, GMT, etc.):

```python
from datetime import datetime

datetime.strptime('Thu, 28 Nov 2024 14:13:56 GMT', "%a, %d %b %Y %H:%M:%S %Z") # throws DTZ007
```

**Comment by jenstroeger** [comment]
More complex string constructions also trigger this:
```python
date_string = ...
date_format = "%Y-%m-%dT%H:%M:%S"
if cond_1:
    ...
    date_format += "%z"
elif cond_2:
    ...
    date_format += "%z"
else:
    date_format += "%z"
datetime.datetime.strptime(date_string, date_format)  # DTZ007
```
